geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderOperationsDUnitTest.java

```diff
@@ -649,6 +677,117 @@ public void testParallelGatewaySenderMessageTooLargeException() {
+
+      doAnswer(invocation -> {
+        // The cache writer is invoked between when the region entry is created with value of
+        // Token.REMOVED_PHASE_1 and when it is replaced with the actual GatewaySenderEvent.
+        // We use this hook to trigger the clear logic when the token is still in the
+        // region entry.
+        cacheWriterLatch.countDown();
+        // Wait until the clear is complete before putting the actual GatewaySenderEvent in the
+        // region entry.
+        shadowRegionClearLatch.await();
+        return null;
+      }).when(mockCacheWriter).beforeCreate(any());
+
+      shadowRegion.setCacheWriter(mockCacheWriter);
+
+      ExecutorService service = Executors.newFixedThreadPool(2);
+
+      List<Callable<Object>> callables = new ArrayList<>();
+
+      // In one thread, we are going to put some test data in the user region,
+      // which will eventually put the GatewaySenderEvent into the shadow region
+      callables.add(Executors.callable(() -> {
+        try {
+          userRegion.put("testKey", "testValue");
+        } catch (Exception ex) {
+          throw new RuntimeException(ex);
+        }
+      }));
+
+      // In another thread, we are clear the shadow region's buckets. If the region entry is a
+      // Token.REMOVED_PHASE_1 at this time, a leak is possible. We can guarantee that the
+      // Token.REMOVED_PHASE_1 is present by using the mocked cache writer defined
+      // above.
+      callables.add(Executors.callable(() -> {
+        try {
+          OffHeapRegionEntryHelper.doWithOffHeapClear(new Runnable() {
+            @Override
+            public void run() {
+              // Wait for the cache writer to be invoked to release this countdown latch.
+              // This guarantees that the region entry will contain a Token.REMOVED_PHASE_1.
+              try {
+                cacheWriterLatch.await();
+              } catch (InterruptedException e) {
+                throw new TestException(
+                    "Thread was interrupted while waiting for mocked cache writer to be invoked");
+              }
+
+              clearShadowBucketRegions(shadowRegion);
+
+              // Signal to the cache writer that the clear is complete and the put can continue.
+              shadowRegionClearLatch.countDown();
+            }
+          });
+        } catch (Exception ex) {
+          throw new RuntimeException(ex);
+        }
+      }));
+
+      List<Future<Object>> futures = service.invokeAll(callables, 10, TimeUnit.SECONDS);
+
+      for (Future<Object> future : futures) {
+        try {
+          future.get();
+        } catch (Exception ex) {
+          throw new TestException(
+              "Exception thrown while executing put and clear concurrently",
+              ex);
+        }
+      }
+
+      userRegion.close();
+
+      Assert.assertEquals("Unexpected orphans found in the cache at end of test", 0,
```

Comment:
Add a Awaitility.await() here as the offheap was released by a background thread.